### ISSUE TITLE
binEnsembleAverage() suggestion

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1201]
I decided to make an issue rather than add more comments to a pull request because I see (open) issues when I go to my oce landing page and not (closed) pull requests.

I've fiddled with `binAverage` #1200 and am thinking it does the right thing. (I could be convinced otherwise!) However, I note from the below that `binEnsembleAverage` could also just use standard R functions instead of having to rely on a `binAverage` that might change.
```R
> i <- 1:25
> split(i, cut(i, breaks=5))
$`(0.976,5.8]`
[1] 1 2 3 4 5

$`(5.8,10.6]`
[1]  6  7  8  9 10

$`(10.6,15.4]`
[1] 11 12 13 14 15

$`(15.4,20.2]`
[1] 16 17 18 19 20

$`(20.2,25]`
[1] 21 22 23 24 25

> as.numeric(lapply(split(i, cut(i, breaks=5)), mean))
[1]  3  8 13 18 23
> 
```

**Comment by dankelley** [comment]
By contrast, in the above, we get
```R
> binAverage(i, i, xinc=5)
$x
[1]  3.5  8.5 13.5 18.5

$y
[1]  3  8 13 18
```

**Comment by dankelley** [comment]
In the `dk` branch (commit b248b5d867815ae4a46a4e8b841023cdadf0a817) I've recoded using `split` and `cut` methodologies, instead of using `binAverage` (see also #1200). Now, I get the number of bins I expected in the output.

I've also added fairly extensive testing at the bottom of `tests/testthat/test_adp.R`; I test the scalars for the first "patch" of averaging, I test a bunch of dimensions, and I test aspects of the `v` matrix (dimension plus one value). I do not test the "raw" matrices, but I'll add that later, likely.  (I just wanted to push to GH before going for a run.)

It could be argued that this `cut` and `split` method is harder to understand, but I'm not actually convinced, because my head is going in circles around `binAverage` (#1200). In the end, this all has to do with the ease of being confused on *increments* versus *number of breaks*. Is my age 2017 minus birth year, or that plus 1 or minus 1? Confuses the heck out of me -- too old to figure it out.

**Comment by dankelley** [comment]
Question for @richardsc -- what should we do if the number of points in the ensemble is not a divisor of the original number of profiles? I think the best scheme would be to first chop off some data at the end, to *make* it a divisor, after which it is simple to follow the user's request directly. Another scheme might be to try to do the best, e.g. letting `cut` give an unequal number of data in the bins [1], but I don't really like that (e.g. then `diff` of the time would not be equal, even if the original data were at a constant separation time interval).

A problem with the above scheme is that we have to discard data *somewhere* and I'm arbitrarily suggesting to do that at the end. Someone might argue that it would be better to do it at the start ... and someone else might say we should let the user decide where. But, for simplicity, I'd prefer to chop at the end, which is a normal ("truncate") sort of approach.

1. Below demonstrates that `cut` will get *all* the data, but that bin lengths won't be equal. My tests seem to show that it is first bin that gets treated differently, but I'm not at all sure on that.
```R
> split(1:11, cut(1:11, breaks=2))
$`(0.99,6]`
[1] 1 2 3 4 5 6

$`(6,11]`
[1]  7  8  9 10 11
```

**Comment by dankelley** [comment]
In `dk` commit f6760d43e4bc00a2679b9f959b3201233d586d0f I now have something that I'm a lot happier with. The test-suite `tests/testthat/test_adp.R` has a test with chopping `data(adp)`, which has 25 profiles, up into 5 profiles. That works as expected. If you try changing to `n<-4` in the test code, you'll see that 7 profiles are created.  The last of these is the leftover one from the original, i.e. the profile that was number 25 in the original.

@richardsc do you like this behaviour of the left-over profile (or profiles) going into a final bin? I feel that I'd prefer to chop that. Otherwise, `diff(adp[["time"]])` will not be constant, which I find odd (and which, I think, might mess up some further analysis unless the user is really paying attention).  Once we get agreement on whether to truncate at the end, I'll tweak the code (including the doc) and I'll add another test.

**Comment by richardsc** [comment]
Honestly, I'm a bit torn about what to do at the end. On the one hand I think the time diffs should be constant. On the other I don't like throwing away potentially a nearly complete ensemble because it is missing 1 ensemble ...

One possibility could be to see what the RDI software does, for guidance. That would at least produce something consistent with what RDI ADCP users are used to. Another thing I think would be useful would be a brief chat with some other "power" users (e.g. Cheel, Hay, etc).

**Comment by dankelley** [comment]
I'll check with RC tomorrow. Also, we could have an arg that controls this, maybe named `discard` or something like that.

**Comment by dankelley** [comment]
I added an arg, and a bunch of tests. So now, people can handle this anyway they like. I made the default to be ignoring extra ensembles, partly because I don't think `plot` on the example code makes it clear that the last ensemble in the resultant value will be at an odd time.

Since I created this issue, I should be the one to close it. But I'll leave this open until tomorrow, in case someone wants to add more discussion.  (I realize I chose a very bad issue title -- according to this title, it's hard to know when to close the issue.)
